design_notes/type_promotion.ipynb

```diff
@@ -0,0 +1,1418 @@
+      "text/plain": [
+       "<Figure size 720x144 with 2 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "#@title\n",
+    "import networkx as nx\n",
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "fig, ax = plt.subplots(1, 2, figsize=(10, 2))\n",
+    "\n",
+    "lattice = {'A': ['B', 'C']}\n",
+    "graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)\n",
+    "pos = {'A': [0, 0], 'B': [1, 0.5], 'C': [1, -0.5]}\n",
+    "nx.draw(graph, with_labels=True, node_size=2000, node_color='lightgray', pos=pos, ax=ax[0], arrowsize=20)\n",
+    "ax[0].set(xlim=[-0.5, 1.5], ylim=[-1, 1])\n",
+    "\n",
+    "lattice = {'A': ['C', 'D'], 'B': ['C', 'D']}\n",
+    "graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)\n",
+    "pos = {'A': [0, 0.5], 'B': [0, -0.5], 'C': [1, 0.5], 'D': [1, -0.5]}\n",
+    "nx.draw(graph, with_labels=True, node_size=2000, node_color='lightgray', pos=pos, ax=ax[1], arrowsize=20)\n",
+    "ax[1].set(xlim=[-0.5, 1.5], ylim=[-1, 1]);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "aDBWlvmezJju"
+   },
+   "source": [
+    "The left DAG is not a lattice because there exists no upper bound for nodes `B` and `C`; the right DAG fails on two counts: first, there exists no upper bound for nodes `C` and `D`, and for nodes `A` and `B` the least upper bound cannot be *uniquely* determined: both `C` and `D` are candidates, but they are unorderable."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "o50FMh8_VGHx"
+   },
+   "source": [
+    "### Properties of a Type Promotion Lattice\n",
+    "\n",
+    "Specifying type promotions in terms of a lattice ensures a number of useful properties. Denoting the lattice join as *J*, we have:\n",
+    "\n",
+    "**Existence:** A lattice by definition requires that a unique lattice join exists for every pair of elements: $\\forall (a, b): \\exists ! J(a, b)$\n",
+    "\n",
+    "**Commutativity:** The lattice join is commutative, i.e. for all *a* and *b* on the lattice, $\\forall (a, b): J(a, b) = J(b, a)$.\n",
+    "\n",
+    "**Associativity:** The lattice join is associative, i.e. for all *a*, *b*, and *c* on the lattice, $\\forall (a, b): J(a, J(b, c)) = J(J(a, b), c)$.\n",
```

Comment:
I think you meant `\forall a, b, c`. I would also suggest using the common symbol `\vee` for join.

Reply:
Good idea - Done